### PR TITLE
Fixed smoke tests

### LINKED ISSUE #1538: Diktat produces wrong `line:column` numbers in the "fix" mode if multiple rules are applied
Diktat **1.2.3**, Ktlint **0.46.1**.
Consider this code fragment:

```kotlin
@file:Suppress(
    "MISSING_KDOC_TOP_LEVEL",
    "MISSING_KDOC_CLASS_ELEMENTS",
    "EMPTY_BLOCK_STRUCTURE_ERROR",
)

package org.cqfn.diktat

class                                        Example{fun memberFunction(){}}
```

Diktat (e.g. when run from the [command line](https://github.com/saveourtool/benedikt/blob/master/docs/diktat-cli.adoc)) reports the following errors, all on the same line:

```console
$ diktat --relative Example.kt
Example.kt:9:6: [TOO_MANY_CONSECUTIVE_SPACES] too many consecutive spaces: found: 40. need to be: 1 (diktat-ruleset:too-many-spaces)
Example.kt:9:53: [WRONG_WHITESPACE] incorrect usage of whitespaces for code separation: there should be a whitespace before '{' (diktat-ruleset:horizontal-whitespace)
Example.kt:9:54: [BRACES_BLOCK_STRUCTURE_ERROR] braces should follow 1TBS style: incorrect same line after opening brace (diktat-ruleset:block-structure)
Example.kt:9:54: [MISSING_KDOC_ON_FUNCTION] all public, internal and protected functions should have Kdoc with proper tags: memberFunction (cannot be auto-corrected) (diktat-ruleset:kdoc-methods)
Example.kt:9:74: [WRONG_WHITESPACE] incorrect usage of whitespaces for code separation: there should be a whitespace before '{' (diktat-ruleset:horizontal-whitespace)
Example.kt:9:76: [BRACES_BLOCK_STRUCTURE_ERROR] braces should follow 1TBS style: no newline before closing brace (diktat-ruleset:block-structure)
```

Now, when run in the "fix" mode, Diktat will correct all of the above errors except for the `MISSING_KDOC_ON_FUNCTION`:

```console
$ diktat --relative -F Example.kt
Example.kt:9:21: [MISSING_KDOC_ON_FUNCTION] all public, internal and protected functions should have Kdoc with proper tags: memberFunction (cannot be auto-corrected) (diktat-ruleset:kdoc-methods)
```

As you can see, the `line:column` numbers will no longer be `9:54` but `9:21`, because some of the rules have been already applied and some of the errors fixed. Yet, `9:21` is absolutely misleading, because the resulting (formatted) file will look like this:

```kotlin
@file:Suppress(
    "MISSING_KDOC_TOP_LEVEL",
    "MISSING_KDOC_CLASS_ELEMENTS",
    "EMPTY_BLOCK_STRUCTURE_ERROR",
)

package org.cqfn.diktat

class Example {
    fun memberFunction() {}
}
```

If we run Diktat again (now against the reformatted file), there will be `10:5` instead of `9:21`:

```console
$ diktat --relative Example.kt
Example.kt:10:5: [MISSING_KDOC_ON_FUNCTION] all public, internal and protected functions should have Kdoc with proper tags: memberFunction (cannot be auto-corrected) (diktat-ruleset:kdoc-methods)
```

The problem becomes even more evident after the upgrade to Ktlint **0.47.x** ([API Changes & RuleSet providers](https://github.com/pinterest/ktlint/releases/tag/0.47.0)), because it has changed the way AST elements are traversed.

Consider there're multiple rules, [ordered](https://github.com/saveourtool/diktat/blob/master/diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/OrderedRuleSet.kt) in a specific way:

```mermaid
graph LR;
    A-->B;
    B-->C;
```

Assuming there're **multiple** AST syntax nodes **enclosed** in each other:
```mermaid
graph TD;
    File-->Node;
    Node-->Leaf;
```

Ktlint **0.46** used to recursively apply the above rules in the following order:

```mermaid
flowchart TB;
  subgraph File
    direction LR;
    A1["A"]-->B1["B"];
    B1["B"]-->C1["C"];
  end
  subgraph Node
    direction LR;
    A2["A"]-->B2["B"];
    B2["B"]-->C2["C"];
  end
  subgraph Leaf
    direction LR;
    A3["A"]-->B3["B"];
    B3["B"]-->C3["C"];
  end
    File-->Node;
    Node-->Leaf;
```

As of Ktlint **0.47**, the traversal order is different:

```mermaid
flowchart TB;
  subgraph A
    direction LR;
    File1["File"]-->Node1["Node"]
    Node1["Node"]-->Leaf1["Leaf"]
  end
  subgraph B
    direction LR;
    File2["File"]-->Node2["Node"]
    Node2["Node"]-->Leaf2["Leaf"]
  end
  subgraph C
    direction LR;
    File3["File"]-->Node3["Node"]
    Node3["Node"]-->Leaf3["Leaf"]
  end
    A-->B;
    B-->C;
```

In practice, when _Diktat_ is run in the "fix" mode, this change in _Ktlint_ behaviour results in the same error (e.g.: `MISSING_KDOC_ON_FUNCTION`) reported multiple times against different (incorrect) locations within a file.

Probably, we should change our strategy to first running `KtLint.format()` ignoring any errors followed by `KtLint.lint()`, and only report errors produced by the last `lint()` call.

In particular, this blocks saveourtool/benedikt#3.

### LINKED ISSUE #1737: Fix tests after migration to ktlint 1.0
All tests marked as `@Disabled("https://github.com/saveourtool/diktat/issues/1737")`

Some of them can fail is due to multiply runs, another -- due to modification of AST tree (limitation of kotlin 1.9):
```
java.lang.IllegalArgumentException: Missing extension point: org.jetbrains.kotlin.com.intellij.treeCopyHandler in container {}
	at org.jetbrains.kotlin.com.intellij.openapi.extensions.impl.ExtensionsAreaImpl.getExtensionPoint(ExtensionsAreaImpl.java:250) ~[kotlin-compiler-embeddable-1.9.10.jar:1.9.10-release-459]
	at org.jetbrains.kotlin.com.intellij.openapi.extensions.BaseExtensionPointName.getPointImpl(BaseExtensionPointName.java:28) ~[kotlin-compiler-embeddable-1.9.10.jar:1.9.10-release-459]
	at org.jetbrains.kotlin.com.intellij.openapi.extensions.ExtensionPointName.getExtensionList(ExtensionPointName.java:39) ~[kotlin-compiler-embeddable-1.9.10.jar:1.9.10-release-459]
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.ChangeUtil.encodeInformation(ChangeUtil.java:43) ~[kotlin-compiler-embeddable-1.9.10.jar:1.9.10-release-459]
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.ChangeUtil.lambda$encodeInformation$0(ChangeUtil.java:39) ~[kotlin-compiler-embeddable-1.9.10.jar:1.9.10-release-459]
	at org.jetbrains.kotlin.com.intellij.psi.impl.DebugUtil.performPsiModification(DebugUtil.java:481) ~[kotlin-compiler-embeddable-1.9.10.jar:1.9.10-release-459]
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.ChangeUtil.encodeInformation(ChangeUtil.java:39) ~[kotlin-compiler-embeddable-1.9.10.jar:1.9.10-release-459]
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.ChangeUtil.copyElement(ChangeUtil.java:95) ~[kotlin-compiler-embeddable-1.9.10.jar:1.9.10-release-459]
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.ChangeUtil.copyElement(ChangeUtil.java:87) ~[kotlin-compiler-embeddable-1.9.10.jar:1.9.10-release-459]
	at org.jetbrains.kotlin.com.intellij.psi.impl.source.tree.TreeElement.copyElement(TreeElement.java:58) ~[kotlin-compiler-embeddable-1.9.10.jar:1.9.10-release-459]
	at com.saveourtool.diktat.ruleset.rules.chapter6.classes.StatelessClassesRule$handleClass$1.invoke(StatelessClassesRule.kt:56) ~[main/:?]
	at com.saveourtool.diktat.ruleset.rules.chapter6.classes.StatelessClassesRule$handleClass$1.invoke(StatelessClassesRule.kt:51) ~[main/:?]
	at com.saveourtool.diktat.ruleset.constants.Warnings.warnAndFix(Warnings.kt:273) ~[main/:?]
```

Possible solutions: https://github.com/pinterest/ktlint/pull/2044

### LINKED ISSUE #1774: Ordering of rules after several re-runs
Workflow of test `ManyLineTransformInLongLineTest`
1. `WhiteSpaceRule` splits to several lines
2. `IndentationRule` adds whitespaces at the start
3. `LineLength` splits again (it's from second run)
4. `IndentationRule` adds whitespaces at the start (it's from second run)

_Originally posted by @nulls in https://github.com/saveourtool/diktat/pull/1769#discussion_r1380057367_

### PR DESCRIPTION
**Pull request by nulls**
### What's done:
- fixed rule PACKAGE_NAME_MISSING

It closes #1737
It closes #1538
It closes #1774

### CODE REVIEW

File: diktat-ruleset/src/test/resources/test/smoke/src/main/kotlin/ManyLineTransformInLongLineExpected.kt [hunk 1]
```diff
@@ -4,7 +4,8 @@ fun foo() {
     (1 or 2 or 3 or 4 or 5 or 6 or 7 or 8 or 9 or 10 or 11 or 12 or 13 or 14 or 15 or 16 or 17 or 18 or 19 or 20 or 21 or 22 or 23 or 24 or 25 or 26 or 27 or 28 or 29 or 30 or 31
     ?: 32 or 33 or 34 or 35 or 36 or 37 or 38 or 39 ?: 40 or 41 or 42 or 43 or 44 or 45 or 46 or 47 or 48 or 49 or 50 or 51 or 52 or 53 + 54 or 55 or 56 or 57 or 58 or 59
     ?: 60 or 61 or 62 or 63 or 64 or 65 or 66 - 67 or 68 or 69 or 70 or 1 + 2 or 3 or 4 or 5 or 6 or 7 or 8 or 9 or 10 or 11 or 12 or 13 or 14 or 15 or 16 or 17 or 18 + 19 -
-        20 or 21 or 22 or 23 or 24 or 25 or 26 or 27 or 28 or 29 or 30 or 31 or 32 or 33 or 34 or 35 or 36 or 37 or 38 or 39 or 40 or 41 || 42 or 43 or 44 or 45 or 46 or 47 && 48 or 49 ||
+        20 or 21 or 22 or 23 or 24 or 25 or 26 or 27 or 28 or 29 or 30 or 31 or 32 or 33 or 34 or 35 or 36 or 37 or 38 or 39 or 40 or 41 || 42 or 43 or 44 or 45 or 46 or 47 &&
+        48 or 49 ||
```
Comment (nulls) [inline diktat-ruleset/src/test/resources/test/smoke/src/main/kotlin/ManyLineTransformInLongLineExpected.kt:8]: it's not corrent and need to fix the rule

File: diktat-rules/src/test/kotlin/com/saveourtool/diktat/ruleset/chapter3/FileStructureRuleFixTest.kt [hunk 2]
```diff
@@ -96,4 +100,50 @@ class FileStructureRuleFixTest : FixTestBase("test/paragraph3/file_structure", :
...
+                    // placeholder
+                }
+            """.trimIndent())
+        }
+        fixAndCompare(
+            expectedFile,
+            testFile,
+            overrideRulesConfigList = listOf(
+                RulesConfig(
+                    name = WarningNames.FILE_NO_BLANK_LINE_BETWEEN_BLOCKS,
```
Comment (nulls) [inline diktat-rules/src/test/kotlin/com/saveourtool/diktat/ruleset/chapter3/FileStructureRuleFixTest.kt:139]: looks like one of `FILE_NO_BLANK_LINE_BETWEEN_BLOCKS` or `WRONG_NEWLINES_AROUND_KDOC` generates two new lines when package is missied. 
Need to analyze PACKAGE_NAME_MISSING -- looks like it puts package in invalid places

**Comment by nulls** [comment]
Can be useful for testing:
```diff
Index: diktat-ktlint-engine/src/main/kotlin/com/saveourtool/diktat/ktlint/KtLintRuleWrapper.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/diktat-ktlint-engine/src/main/kotlin/com/saveourtool/diktat/ktlint/KtLintRuleWrapper.kt b/diktat-ktlint-engine/src/main/kotlin/com/saveourtool/diktat/ktlint/KtLintRuleWrapper.kt
--- a/diktat-ktlint-engine/src/main/kotlin/com/saveourtool/diktat/ktlint/KtLintRuleWrapper.kt	(revision 39de04c9bc487b601dd1081f70d1843c974fe558)
+++ b/diktat-ktlint-engine/src/main/kotlin/com/saveourtool/diktat/ktlint/KtLintRuleWrapper.kt	(date 1698917676540)
@@ -7,7 +7,12 @@
 import com.pinterest.ktlint.rule.engine.core.api.Rule.VisitorModifier.RunAfterRule.Mode
 import com.pinterest.ktlint.rule.engine.core.api.RuleId
 import com.pinterest.ktlint.rule.engine.core.api.RuleProvider
+import com.pinterest.ktlint.rule.engine.core.api.editorconfig.EditorConfig
 import org.jetbrains.kotlin.com.intellij.lang.ASTNode
+import java.nio.file.Paths
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+import kotlin.io.path.writeText
 
 private typealias EmitType = (offset: Int, errorMessage: String, canBeAutoCorrected: Boolean) -> Unit
 
@@ -23,15 +28,34 @@
     about = about,
     visitorModifiers = createVisitorModifiers(rule, prevRuleId),
 ) {
+    override fun afterLastNode() {
+        visited.set(false)
+    }
+
     override fun beforeVisitChildNodes(
         node: ASTNode,
         autoCorrect: Boolean,
         emit: EmitType,
-    ) = rule.invoke(node, autoCorrect) { offset, errorMessage, canBeAutoCorrected ->
-        emit.invoke(offset, errorMessage.correctErrorDetail(canBeAutoCorrected), canBeAutoCorrected)
+    ) {
+        val folder = Paths.get("d:\\projects\\diktat\\test")
+        val doPrint = visited.compareAndSet(false, true)
+        if (doPrint) {
+            folder.resolve(counter.getAndIncrement().toString().removePrefix("1") + "_before_" + rule.javaClass.simpleName + ".txt")
+                .writeText(node.prettyPrint())
+        }
+        rule.invoke(node, autoCorrect) { offset, errorMessage, canBeAutoCorrected ->
+            emit.invoke(offset, errorMessage.correctErrorDetail(canBeAutoCorrected), canBeAutoCorrected)
+        }
+        if (doPrint) {
+            folder.resolve(counter.getAndIncrement().toString().removePrefix("1") + "_after_" + rule.javaClass.simpleName + ".txt")
+                .writeText(node.prettyPrint())
+        }
     }
 
     companion object {
+        private val counter = AtomicInteger(1000)
+        private val visited = AtomicBoolean(false)
+
         private val about: About = About(
             maintainer = "Diktat",
             repositoryUrl = "https://github.com/saveourtool/diktat",
@@ -73,5 +97,29 @@
          * @return a rule to which a logic is delegated
          */
         internal fun Rule.unwrap(): DiktatRule = (this as? KtLintRuleWrapper)?.rule ?: error("Provided rule ${javaClass.simpleName} is not wrapped by diktat")
+
+
+        /**
+         * Converts this AST node and all its children to pretty string representation
+         */
+        @Suppress("AVOID_NESTED_FUNCTIONS")
+        fun ASTNode.prettyPrint(level: Int = 0, maxLevel: Int = -1): String {
+            /**
+             * AST operates with \n only, so we need to build the whole string representation and then change line separator
+             */
+            fun ASTNode.doPrettyPrint(level: Int, maxLevel: Int): String {
+                val result = StringBuilder("${this.elementType}: \"${this.text}\"").append('\n')
+                if (maxLevel != 0) {
+                    this.getChildren(null).forEach { child ->
+                        result.append(
+                            "${"-".repeat(level + 1)} " +
+                                    child.doPrettyPrint(level + 1, maxLevel - 1)
+                        )
+                    }
+                }
+                return result.toString()
+            }
+            return doPrettyPrint(level, maxLevel).replace("\n", System.lineSeparator())
+        }
     }
 }

```

File: diktat-rules/src/test/kotlin/com/saveourtool/diktat/ruleset/chapter3/FileStructureRuleFixTest.kt (hunk 2 above)
Comment (nulls) [reply diktat-rules/src/test/kotlin/com/saveourtool/diktat/ruleset/chapter3/FileStructureRuleFixTest.kt]: the problem was in empty import list -- PACKAGE rule adds the WHITE_SPACE after package, but FILE_STRUCTURE rule adds it after IMPORT_LIST

File: diktat-ruleset/src/test/resources/test/smoke/src/main/kotlin/ManyLineTransformInLongLineExpected.kt (hunk 1 above)
Comment (nulls) [reply diktat-ruleset/src/test/resources/test/smoke/src/main/kotlin/ManyLineTransformInLongLineExpected.kt:8]: looks like it's by design:
# WhiteSpaceRule splits to several lines
# IndentationRule adds whitespaces at the start
# LineLength splits again (it's from second run)
# IndentationRule adds whitespaces at the start (it's from second run)

File: diktat-rules/src/main/kotlin/com/saveourtool/diktat/ruleset/rules/DiktatRuleSetFactoryImpl.kt [hunk 3]
```diff
@@ -117,9 +117,9 @@ class DiktatRuleSetFactoryImpl : DiktatRuleSetFactory {
             // comments & documentation
             ::CommentsRule,
             ::SingleConstructorRule,  // this rule can add properties to a primary constructor, so should be before KdocComments
-            ::KdocComments,
             ::KdocMethods,
             ::KdocFormatting,
+            ::KdocComments,
```
Comment (nulls) [inline diktat-rules/src/main/kotlin/com/saveourtool/diktat/ruleset/rules/DiktatRuleSetFactoryImpl.kt:122]: the problem is not in ordering:
1. We check KDoc when process `CLASS`
2. We fix the KDoc when process `VAL_KEYWORD` or `VAR_KEYWORD` inside.

We can move checking KDoc on CLASS level after processing the childs. But probably the fixing of KDoc for class should be done on class level

Comment (nulls) [reply diktat-rules/src/main/kotlin/com/saveourtool/diktat/ruleset/rules/DiktatRuleSetFactoryImpl.kt]: created an issue for it #1776

**Review by akuleshov7 — APPROVED** [review]
tests, passed, good

File: diktat-ruleset/src/test/resources/test/smoke/src/main/kotlin/ManyLineTransformInLongLineExpected.kt (hunk 1 above)
Comment (TWiStErRob) [reply diktat-ruleset/src/test/resources/test/smoke/src/main/kotlin/ManyLineTransformInLongLineExpected.kt:8]: I love this test!

**Pull request merged**